File: Sources/CryptoMakoS3/ObjectStore.swift
import Foundation

/// S3-backed blob store used by the vault layer.
public protocol ObjectStore: Sendable {
    func getObject(key: String) async throws -> Data
    func getObject(key: String, to fileURL: URL) async throws
    func listImmediate(prefix: String) async throws -> PrefixListing
}

public struct PrefixListing: Sendable {
    public var objects: [ListedObject]
    public var commonPrefixes: [String]

    public init(objects: [ListedObject] = [], commonPrefixes: [String] = []) {
        self.objects = objects
        self.commonPrefixes = commonPrefixes
    }
}

public struct ListedObject: Sendable {
    public var key: String
    public var size: Int64
    public var eTag: String?

    public init(key: String, size: Int64, eTag: String? = nil) {
        self.key = key
        self.size = size
        self.eTag = eTag
    }
}

public enum ObjectStoreError: Error, LocalizedError {
    case notFound(String)
    case transport(String)

    public var errorDescription: String? {
        switch self {
        case .notFound(let key):
            return "object not found: \(key)"
        case .transport(let message):
            return message
        }
    }
}

File: windows/src/CryptoMako.Vault/ObjectStore.cs
namespace CryptoMako.Vault;

/// <summary>
/// Blob store used by the vault layer. Durable success = remote put/delete ACK on S3;
/// <see cref="DirectoryObjectStore"/> is for --local / tests only.
/// </summary>
public interface IObjectStore
{
    Task<byte[]> GetObjectAsync(string key, CancellationToken ct = default);
    Task GetObjectAsync(string key, string destinationPath, CancellationToken ct = default);
    Task<ListedObject> HeadObjectAsync(string key, CancellationToken ct = default);
    Task<PrefixListing> ListImmediateAsync(string prefix, CancellationToken ct = default);
    Task PutObjectAsync(string key, byte[] data, CancellationToken ct = default);
    Task PutObjectAsync(string key, string sourceFilePath, CancellationToken ct = default);
    Task DeleteObjectAsync(string key, CancellationToken ct = default);
}

public sealed class PrefixListing
{
    public List<ListedObject> Objects { get; init; } = new();
    public List<string> CommonPrefixes { get; init; } = new();
}

public sealed class ListedObject
{
    public required string Key { get; init; }
    public long Size { get; init; }
    public string? ETag { get; init; }
}

public class ObjectStoreException : Exception
{
    public ObjectStoreException(string message) : base(message) { }
    public ObjectStoreException(string message, Exception inner) : base(message, inner) { }
}

public sealed class ObjectNotFoundException : ObjectStoreException
{
    public string Key { get; }
    public ObjectNotFoundException(string key) : base($"object not found: {key}") => Key = key;
}
